op-node/rollup/derive/l1_retrieval.go

```diff
@@ -18,69 +17,68 @@ type L1Retrieval struct {
 	log     log.Logger
 	dataSrc *CalldataSource
 	next    L1SourceOutput
+	prev    *L1Traversal
 
 	progress Progress
 
-	data  eth.Data
 	datas *CalldataSourceImpl
 }
 
 var _ Stage = (*L1Retrieval)(nil)
 
-func NewL1Retrieval(log log.Logger, dataSrc *CalldataSource, next L1SourceOutput) *L1Retrieval {
+func NewL1Retrieval(log log.Logger, dataSrc *CalldataSource, next L1SourceOutput, prev *L1Traversal) *L1Retrieval {
 	return &L1Retrieval{
 		log:     log,
 		dataSrc: dataSrc,
 		next:    next,
+		prev:    prev,
 	}
 }
 
 func (l1r *L1Retrieval) Progress() Progress {
 	return l1r.progress
 }
 
-func (l1r *L1Retrieval) Step(ctx context.Context, outer Progress) error {
-	if changed, err := l1r.progress.Update(outer); err != nil || changed {
-		return err
-	}
-
-	// specific to L1 source: if the L1 origin is closed, there is no more data to retrieve.
-	if l1r.progress.Closed {
-		return io.EOF
-	}
-
-	// create a source if we have none
-	if l1r.datas == nil {
-		l1r.datas = l1r.dataSrc.OpenData(ctx, l1r.progress.Origin.ID())
-		return nil
-	}
-
-	// buffer data if we have none
-	if l1r.data == nil {
+// Step does an action in the L1 Retrieval stage
+// If there is data, it pushes it to the next stage.
+// If there is no more data open ourselves if we are closed or close ourselves if we are open
+func (l1r *L1Retrieval) Step(ctx context.Context, _ Progress) error {
+	if l1r.datas != nil {
 		l1r.log.Debug("fetching next piece of data")
 		data, err := l1r.datas.Next(ctx)
 		if err == io.EOF {
-			l1r.progress.Closed = true
 			l1r.datas = nil
 			return io.EOF
 		} else if err != nil {
 			return NewTemporaryError(fmt.Errorf("context to retrieve next L1 data failed: %w", err))
 		} else {
-			l1r.data = data
+			l1r.next.IngestData(data)
 			return nil
```

Comment:
I would love it better if stages where to return `(data, err)` instead of pushing directly into the next stage.
Ideally, a stage wouldn't ever be passed its next stage (only relevant input from it).